pkg/solana/logpoller/loader_test.go

```diff
@@ -53,42 +58,178 @@ func TestEncodedLogCollector_ParseSingleEvent(t *testing.T) {
 		require.NoError(t, collector.Close())
 	})
 
-	slot := uint64(42)
-	sig := solana.Signature{2, 1, 4, 2}
-	blockHeight := uint64(21)
+	var latest atomic.Uint64
 
-	client.EXPECT().GetLatestBlockhash(mock.Anything, rpc.CommitmentFinalized).Return(&rpc.GetLatestBlockhashResult{
-		RPCContext: rpc.RPCContext{
-			Context: rpc.Context{
-				Slot: slot,
-			},
-		},
-	}, nil)
+	latest.Store(uint64(40))
 
-	client.EXPECT().GetBlocks(mock.Anything, uint64(1), mock.MatchedBy(func(val *uint64) bool {
-		return val != nil && *val == slot
-	}), mock.Anything).Return(rpc.BlocksResult{slot}, nil)
+	client.EXPECT().
+		GetLatestBlockhash(mock.Anything, rpc.CommitmentFinalized).
+		RunAndReturn(latestBlockhashReturnFunc(&latest))
 
-	client.EXPECT().GetBlockWithOpts(mock.Anything, slot, mock.Anything).Return(&rpc.GetBlockResult{
-		Transactions: []rpc.TransactionWithMeta{
-			{
-				Meta: &rpc.TransactionMeta{
-					LogMessages: messages,
-				},
-			},
-		},
-		Signatures:  []solana.Signature{sig},
-		BlockHeight: &blockHeight,
-	}, nil).Twice()
+	client.EXPECT().
+		GetBlocks(
+			mock.Anything,
+			mock.MatchedBy(getBlocksStartValMatcher),
+			mock.MatchedBy(getBlocksEndValMatcher(&latest)),
+			rpc.CommitmentFinalized,
+		).
+		RunAndReturn(getBlocksReturnFunc(false))
+
+	client.EXPECT().
+		GetBlockWithOpts(mock.Anything, mock.Anything, mock.Anything).
+		RunAndReturn(func(_ context.Context, slot uint64, _ *rpc.GetBlockOpts) (*rpc.GetBlockResult, error) {
+			height := slot - 1
+
+			result := rpc.GetBlockResult{
+				Transactions: []rpc.TransactionWithMeta{},
+				Signatures:   []solana.Signature{},
+				BlockHeight:  &height,
+			}
+
+			_, _ = rand.Read(result.Blockhash[:])
+
+			if slot == 42 {
+				var sig solana.Signature
+				_, _ = rand.Read(sig[:])
+
+				result.Signatures = []solana.Signature{sig}
+				result.Transactions = []rpc.TransactionWithMeta{
+					{
+						Meta: &rpc.TransactionMeta{
+							LogMessages: messages,
+						},
+					},
+				}
+			}
+
+			return &result, nil
+		})
 
 	tests.AssertEventually(t, func() bool {
 		return parser.Called()
 	})
+}
+
+func TestEncodedLogCollector_MultipleEventOrdered(t *testing.T) {
```

Comment:
I think this test only passes because it's testing only 2 slots. For a sequence of 3 or more slots I think it should fail with the existing code.

The reason 2 slots passes is because the first element in the heap is always the first element in the slice backing it. But beyond that I think it depends on the order in which elements were added. Can we make it test 3 or more slots, to catch this regression?

Reply:
Hmm... I think it must still be a false positive for some reason, I'm pretty sure the code as written does not guarantee the order.

Just ran the debugger on this test, and I notice that the elements are always getting added to `p.blocks` in ascending order (41, 42, 43, 44).  If that's guaranteed to happen in production, then we can probably just go back to having it be an ordinary slice rather than a heap, and don't need to sort it.  But I suspect there is a way it can happen--we just need to find a way of testing it. Maybe it would happen if there are two different requests for `GetBlocks()` sent and those came back out of order?  I see that you added a 1s delay in the mocked `GetBlockWithOpts` request... but I think that doesn't make a difference here, because `ExpectBlock()` is called before that's ever called.

Reply:
I added a more granular test `TestBlockHeap` to check that out of order elements added are iterated over in order:
https://github.com/smartcontractkit/chainlink-solana/pull/975

It does fail for ordinary iteration and succeeds when using `Pop`

Reply:
I think we should add this test and fix the iteration, then comment out the part of the test that's failing.

Would be nice to also have a test where multiple calls to `GetBlocks` come back in the wrong order, but maybe this is enough?

Reply:
I agree with Domino that iteration over the heap's slice does not guarantee order. 
Also I do not see why we need heap or sorting? Do we expect that `GetBlocks` may return blocks out-of-order? It seems like linked should be enough.